### ISSUE TITLE
ignore_index does not have expected behavior

### ISSUE THREAD

**Issue by DeeDive** [pytorch/pytorch#25166]
Recently I've encountered a problem and find that ignore_index  in CrossEntropyLoss does not have expected behavior. I want to completely ignore samples with a particular label when calculating loss.

my problem is as follows:
I have a mixture of Labelled Sample Group and Unlabeled Sample Group. For some special reason, I have to mix those samples in a same mini-batch to train a classification network, with CrossEntropyLoss . So I need to ignore these unlabeled samples when computing the CrossEntropyLoss.

I tried to assign a pseudo label to these unlabeled samples and tried to set the option weight in CrossEntropyLoss to be ( zero ) or set the ignore_index corresponding to the pseudo label  , but found that both wouldn’t work properly. Although the loss on the pseudo class is ignored, if the prediction of an unlabeled data have large values on other indexes, it still generate large loss. I need those samples contribute exactly 0 to the CrossEntropyLoss.

So I think the behavior of ignore_index is not proper,isn't it?

Is there something wrong with me and will someone kindly help me with this issue?
Millions of thanks in advance.

**Comment by andreaskoepf** [comment]
Could you please explain what you mean by: "(...) if the prediction of an unlabeled data have large values on other indexes, it still generate large loss."?

If `ignore_index` is supplied `CrossEntropyLoss` will ignore all batch-entries having a target-class equal to `ignore_index`, e.g. a simple test:

```
x = torch.randn(10, 10)
y = torch.zeros(10, dtype=torch.long)
c = nn.CrossEntropyLoss(ignore_index=0)
c(x, y)
````
shows `tensor(0.)`.

Could you show a minimal example script that shows this would not be the case?

**Comment by Infinitestudy** [comment]
Hi @andreaskoepf ,
Sorry for the wrong description, actually I mean if labeled data was misclassified to the ignored label.
For example, 
```
>>> x
tensor([[-1.8799,  0.2902, -0.7727],
        [ 0.9352,  0.3834,  0.2513]])
>>> y=torch.toTensor([0,1])
>>> c = nn.CrossEntropyLoss(ignore_index=0)
>>> c(x,y)
tensor(1.2845)
```
here, the second sample is misclassified to label 0, which lead to the final loss 1.2845.
So, the label specified by variable ignore_index still contributes to the computation of softmax.

I personally think ignore_index should completely remove any effect caused by that label, which will contribute 0 to the loss.

**Comment by andreaskoepf** [comment]
@Infinitestudy thanks for clarifying the problem. 

A quick-fix would be to write `-inf` into the first column, e.g. `x[:,0] = float('-inf')`.

But I think you could better specify  an `ignore_index` that lies outside the class range for the outputs you want to ignore. 

You net should only have output values for real classes.

```
c = nn.CrossEntropyLoss(ignore_index=-1)
x = torch.rand(2, 10)
y = torch.tensor([-1, 4])
c(x, y)     # will ignore row 0 since target class is -1
```

**Comment by mruberry** [comment]
ignore_index does appear to be confusing: see also https://github.com/pytorch/pytorch/issues/18206. However it appears this issue has been addressed, so I'm going to close it. But please reopen if you'd still like to see something addressed!
